### PR TITLE
Update Covenant Descriptor module

### PR DESCRIPTION
**Pull request by sanket1729**
Also fixes bugs in covenant descriptor construction

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -95,6 +95,8 @@
 //!
 //!
 #![allow(bare_trait_objects)]
+// Required for rustc 1.29
+#![recursion_limit = "128"]
```
Comment (sanket1729) [inline src/lib.rs:99]: I don't know how serious is this issue. In the next commit, our current implementation of match_token requires this. It is possible to get around this, but in the interest of time I have decided to ignore this

Comment (apoelstra) [reply src/lib.rs:99]: lol!! Looks good to me.

For future reference this is because of the massive `match_tokens!` invocation in cov.rs

**Review by apoelstra — APPROVED** [review]
ack 3230ebb84e4b7e0a5eb08d47eddfc8297f45c652

Notes
* this uses a one-off branch of rust-elements which eventually we'll need to fix
* it comments out the PSET unit tests which we'll eventually need to fix
* I'm not thrilled with all the repetitive size checks but I don't see any way around them :|

**Comment by apoelstra** [comment]
ack ccb82e5b6b8e055f9cc0173cc738f4c0738310b4 I mean

I ran `git rev-parse pr/9/head` on the wrong repo :)

**Comment by sanket1729** [comment]
This had to redone because of upstream changes in rust-elements. Could not build on top of ccb82e5, had to rebase from the first commit

File: examples/sign_multisig.rs [hunk 2]
```diff
@@ -18,7 +18,7 @@ extern crate bitcoin;
 extern crate elements;
 extern crate elements_miniscript as miniscript;
 
-use bitcoin::secp256k1; // secp256k1 re-exported from rust-bitcoin
+use elements::secp256k1_zkp; // secp256k1 re-exported from rust-bitcoin
```
Comment (apoelstra) [inline examples/sign_multisig.rs:21]: should remove this comment

File: src/pset/finalizer.rs [hunk 3]
```diff
@@ -204,6 +204,53 @@ pub(super) fn get_descriptor(
...
+) -> Result<(), Error> {
+    let cltv = pset
+        .locktime()
+        .map_err(|_e| Error::LockTimeCombinationError)?;
+    let input = &pset.inputs[index];
+
+    let spk = get_scriptpubkey(pset, index).map_err(|e| Error::InputError(e, index))?;
+    let empty_script_sig = Script::new();
+    let empty_witness = Vec::new();
+    let script_sig = input.final_script_sig.as_ref().unwrap_or(&empty_script_sig);
```
Comment (apoelstra) [inline src/pset/finalizer.rs:222]: These `empty_script_sig` and `empty_witness` variables feel a bit weird. It's zero-cost to produce an empty `Script` or `Vec`, you can just do it inline.

**Review by apoelstra — APPROVED** [review]
ack dcc606ff329b2b95f6bedaa9c2376c3202f52181

Though my previous comments (in particular that we need a new rust-elements release) still apply; also there are a few spelling errors in the comments.

File: src/pset/finalizer.rs (hunk 3 above)
Comment (sanket1729) [reply src/pset/finalizer.rs:222]: I think we need a reference to a Script or witness and I cannot do create Script inline and get a reference to it inline.

**Comment by sanket1729** [comment]
I have added a spell check to my IDE, hopefully, there won't be very many spell errors in the future.

**Comment by sanket1729** [comment]
Will rebase once we make a release of rust-elements

**Review by apoelstra — APPROVED** [review]
ack 98b96cae34ea474bce4cd695969837f700108d77

**Pull request merged**